### ISSUE TITLE
Add support for `tmux` Control Mode

### ISSUE THREAD

**Issue by zadjii-msft** [microsoft/terminal#3656]
From discussion in #532, #1000:

> If we're looking at tmux for things like splits, why not implement [Tmux Control Mode](https://medium.com/@gveloper/using-iterm2s-built-in-integration-with-tmux-d5d0ef55ec30) like iTerm on MacOS. I switch back and forth between my Macs (with iTerm2) and my Surface Laptop with WSL but WSL2 combined with a _really_ usable terminal (comparable to iTerm2) would really make me consider leaving the Apple camp.
> 

> In #532, from @zadjii-msft:
> 
> > I did not know that Tmux Control Mode even existed. That sounds really awesome, I'm totally on board with trying to make that work somehow.
> 
> Was there any further thought put into this? I love tmux control mode - it's literally the only thing keeping me tethered to MacOS at work (since only iTerm supports it, and iTerm is mac-only). If it were implemented, I'd trash my Mac and request a Windows laptop immediately.
> 
> I'd love to see it listed as a post-1.0 milestone :)

We should do this.

**Comment by nicm** [comment]
Control mode isn't documented all that well but it isn't terribly complicated, if you want to do it you can feel free to ask me if you have questions.

**Comment by DHowett-MSFT** [comment]
@nicm Thanks! Maybe we can contribute a bit to the existence of control mode docs while we're at it. :smile:

**Comment by nicm** [comment]
That would be great. There is an overview in the manual and it is easy enough to play with interactively using "tmux -C new" (instead of -CC). Plus iTerm2 has an implementation of the client side of course.

**Comment by ArcticLight** [comment]
I was actually just looking at Tmux Control Mode for work (which led me to this issue from Google) - this would be something I would really support! That, and there are already library implementations which might help; as @nicm mentioned iTerm2 implements control mode, and also I found this python library https://libtmux.git-pull.com/en/latest/ which implements the control mode client, too.

**Comment by nicm** [comment]
I don't think libtmux uses control mode.

**Comment by nicm** [comment]
I have written a wiki entry about control mode, although it doesn't have an awful lot more than is in the manual at the moment. Still suggestions welcome - https://github.com/tmux/tmux/wiki/Control_Mode.

**Comment by muodov** [comment]
I would love to have this feature - it's one of the crucial things I miss from my MacOS days.
For those who end up here - push that Like button above! Let's make ourselves heard, surely the product team will notice this! :D

**Comment by skyline75489** [comment]
I think this depends heavily on the remoting abiliby in #7240 & #8135. A tmux window will have to be a different one than the window that executes `tmux -CC`. And they must be able to communicate with each other.

**Comment by joexue** [comment]
One thing that prevent me from moving from Mac to Windows is the iTerm tmux control mode

**Comment by skyline75489** [comment]
See also https://github.com/wez/wezterm/pull/1090

**Comment by skyline75489** [comment]
Update: part of this PR https://github.com/wez/wezterm/pull/1090 has been merged into main branch. But I think on Windows the tmux-cc feature in wezterm is still blocked by ConPTY. Haven't got the chance to try it, though.

**Comment by jayPare** [comment]
This would be so helpful for me! Any update/progress?

**Comment by zadjii-msft** [comment]
Nope. We'll make sure to update this thread when there is. In the meantime, might I recommend the Subscribe button?
![image](https://user-images.githubusercontent.com/18356694/91237459-5cbb0c80-e700-11ea-9347-b9b1ec2813b1.png)
That way you'll be notified of any updates to this thread, without needlessly pinging everyone on this thread ☺️

**Comment by joexue** [comment]
> 

+1

**Comment by joexue** [comment]
> This would be so helpful for me! Any update/progress?

+1

**Comment by nrclark** [comment]
+1 for tmux control mode!

This is implemented on iTerm2 for MacOS, and it's crazy magic that gives a native feeling to working remotely over SSH. It's also nice because I can disconnect/reconnect without losing all of my tabs and whatnot.

If Windows Terminal learned how to use control-mode, I'd switch my work computer over in an instant. And I don't think I'm the only one.

**Comment by shinebayar-g** [comment]
It'd make so much sense since Windows Terminal already supports split panes, it'd fit perfectly <3 please support this @Microsoft

**Comment by BinaryShrub** [comment]
Any update on this? I use iterm2 exclusively because of this integration

**Comment by iDarshan** [comment]
Any update on this? This is the most wanted feature!

**Comment by susanthenerd** [comment]
I would also like to see this happen. Any update?

**Comment by rogerday** [comment]
Agreed with everyone this is absolutely a killer feature in iTerm, once you've got used to it it's __really__ annoying to have to use another terminal.

**Comment by jayPare** [comment]
There is an ongoing PR to support Control Mode for wezterm here:
https://github.com/wez/wezterm/pull/6602

I didn't looked at the code since I'm not very familiar with Wezterm code base, but I built everything from source on Windows and tested connecting to tmux 3.4 on a Linux machine. Everything seems to work properly. If anyone is a little familiar with wezterm code base, please go ahead and review the PR.

For those that want to test it, you can build from sources  using the following steps:
```
git clone https://github.com/wez/wezterm.git
git pull origin pull/6602/head --rebase
git submodule init
git submodule update
```
On Windows, you can follow the step to build from source from [here](https://wezfurlong.org/wezterm/install/source.html#installing-from-source):

1. Install [Strawberry Perl](https://strawberryperl.com/).
2. Do the following command to add Strawberry Perl to you path:
`set PATH=c:\Strawberry\perl\bin;%PATH%`
3. Install Visual Studio 2022 Community Edition from [here](https://visualstudio.microsoft.com/downloads/)
4. Configure it for Desktop development with C++
5. Install MSVC version of Rust from [here](https://forge.rust-lang.org/infra/other-installation-methods.html) (most likely x86_64-pc-windows-msvc)
6. Use cargo to build wezterm from source:
`cargo build --release`
7. Run WezTerm with the following line:
`cargo run --release --bin wezterm`

If you encounter any issue, or if I forgot some steps, please let me know.

**Comment by joexue** [comment]
A PR #18928 for this feature is raised.
Please go to comment, review, test and criticize.

![Screenshot 2025-05-13 232209](https://github.com/user-attachments/assets/a65dfaa2-55b1-4a37-a1fa-d887cd4b8531)

**Comment by nrclark** [comment]
@joexue fantastic!! Thanks for your hard work.
